tianshou/data/collector.py

```diff
@@ -34,13 +34,16 @@ class Collector(object):
         with corresponding policy's exploration noise. If so, "policy.
         exploration_noise(act, batch)" will be called automatically to add the
         exploration noise into action. Default to False.
+    :param bool gym_reset_return_info: if set to True, return the info dict when
```

Comment:
Can we do the same things for collector? I mean dynamic checking instead of using user-defined flags (because the user's assumption is, `collector(policy, env, buffer)` will work in any way, no need to pass in other args in most use cases). If so, it will pop up tons of open issues when gym release 1.0, shift to return `(obs, info)` by default and delete the `return_info` flag.

Reply:
In other places, doing dynamic checking made sense because we allow users to specify arbitrary kwargs when calling `reset`, since user can specify `return_info=True`. For `Collector.collect`, what do you think is the best way for users to specify `return_info`? I'm thinking either have `gym_reset_kwargs` as an input arg in `Collector.collect` and `Collector.reset`. Then we can dynamically check the output.

Reply:
I mean you can add an end-to-end training test with `envpool.make_gym(gym_reset_return_info=True)`...

Reply:
supporting `envpool.make_gym(gym_reset_return_info=True)` through dynamic checking sounds good and I can also add an end-to-end test. My question was more geared towards supporting `return_info` for non-envpool environments where you pass this argument in through the reset function, should the collector support this? (i.e. `collector.collect(gym_reset_kwargs={return_info=True})`)

Reply:
Sure, I think we should do that.

Reply:
done in https://github.com/thu-ml/tianshou/pull/613/commits/c2ff71fb5ae1a5dae99dd677e3e72a274b28566a
Although one issue I ran into is that I had to do away with your earlier [suggestion](https://github.com/thu-ml/tianshou/pull/613#discussion_r878699042) of caching the `has_info` result, because [collector's __init__ function](https://github.com/thu-ml/tianshou/blob/master/tianshou/data/collector.py#L77) calls `self.reset` without any kwargs, so we end up always caching "not returning info". Wondering if you have any better ideas. I think maybe it's ok to check on every reset since it only takes a couple of microseconds on my computer and after gym version 0.5, the API is changing to always returning info.

Reply:
Instead of an end-to-end test with `envpool.make_gym(gym_reset_return_info=True)`, I added a collector unit test that uses envpool, see: `test_collector_envpool_gym_reset_return_info`

Reply:
> Wondering if you have any better ideas.

I think the current solution is fine, but if I were you, again I'll implement option 3 in https://github.com/thu-ml/tianshou/pull/613#discussion_r878787085 because that's cleaner.

Reply:
One other problem that I forsee with option 2 is: envpool and gym environments return (obs, info) through different ways (envpool through `gym_reset_return_info` when initializing the env, and gym through `return_info` when calling reset), so the venvs needs to be smart about whether `return_info` can be included as a kwarg when calling `reset`. we might have to do something like:
```
if not hasattr(self.env, `gym_reset_return_info`):
    obs, info = self.env.reset(return_info=True, **kwargs)
elif not self.env.gym_reset_return_info:
    raise ValueError("gym_reset_return_info must be True for envpool environments")
else:
    obs, info = self.env.reset(**kwargs)
```
if you think this is still the way to go, I can implement it instead.

Reply:
And let me know if there are other issues or if we are good to merge

Reply:
Just returned from vacation. I'll review it as soon as possible. Sorry for the delay.

Reply:
@Trinkle23897 can you give this PR another review?